chips/lowrisc/src/spi_host.rs

```diff
@@ -0,0 +1,672 @@
+        CLKDIV_0 OFFSET(0) NUMBITS(16) [],
+        CSNIDLE_0 OFFSET(16) NUMBITS(3) [],
+        CSNTRAIL_0 OFFSET(20) NUMBITS(3) [],
+        CSNLEAD_0 OFFSET(24) NUMBITS(3) [],
+        //28 Reserved
+        FULLCYC_0 OFFSET(29) NUMBITS(1) [],
+        CPHA_0 OFFSET(30) NUMBITS(1) [],
+        CPOL_0 OFFSET(31) NUMBITS(1) [],
+    ],
+    csid_ctrl [
+        CSID OFFSET(0) NUMBITS(32) [],
+    ],
+    command [
+        LEN OFFSET(0) NUMBITS(8) [],
+        CSAAT OFFSET(9) NUMBITS(1) [],
+        SPEED OFFSET(10) NUMBITS(2) [],
+        DIRECTION OFFSET(12) NUMBITS(2) [],
+    ],
+    rx_data [
+        DATA OFFSET(0) NUMBITS(32) [],
+    ],
+    tx_data [
+        DATA OFFSET(0) NUMBITS(32) [],
+    ],
+    err_en [
+        CMDBUSY OFFSET(0) NUMBITS(1) [],
+        OVERFLOW OFFSET(1) NUMBITS(1) [],
+        UNDERFLOW OFFSET(2) NUMBITS(1) [],
+        CMDINVAL OFFSET(3) NUMBITS(1) [],
+        CSIDINVAL OFFSET(4) NUMBITS(1) [],
+    ],
+    err_status [
+        CMDBUSY OFFSET(0) NUMBITS(1) [],
+        OVERFLOW OFFSET(1) NUMBITS(1) [],
+        UNDERFLOW OFFSET(2) NUMBITS(1) [],
+        CMDINVAL OFFSET(3) NUMBITS(1) [],
+        CSIDINVAL OFFSET(4) NUMBITS(1) [],
+        ACCESSINVAL OFFSET(5) NUMBITS(1) [],
+    ],
+    event_en [
+        RXFULL OFFSET(0) NUMBITS(1) [],
+        TXEMPTY OFFSET(1) NUMBITS(1) [],
+        RXWM OFFSET(2) NUMBITS(1) [],
+        TXWM OFFSET(3) NUMBITS(1) [],
+        READY OFFSET(4) NUMBITS(1) [],
+        IDLE OFFSET(5) NUMBITS(1) [],
+    ],
+];
+
+pub struct SpiHost {
+    registers: StaticRef<SpiHostRegisters>,
+    client: OptionalCell<&'static dyn hil::spi::SpiMasterClient>,
+    busy: Cell<bool>,
+    chip_select: Cell<u32>,
+    cpu_clk: Cell<u32>,
+    tsclk: Cell<u32>,
+    tx_buf: TakeCell<'static, [u8]>,
+    rx_buf: TakeCell<'static, [u8]>,
+    tx_len: Cell<usize>,
+    rx_len: Cell<usize>,
+    tx_offset: Cell<usize>,
+    rx_offset: Cell<usize>,
+}
+// SPI Host Command Direction: Bidirectional
+const SPI_HOST_CMD_BIDIRECTIONAL: u32 = 3;
+// SPI Host Command Speed: Standard SPI
+const SPI_HOST_CMD_STANDARD_SPI: u32 = 0;
+
+impl SpiHost {
+    pub const fn new(base: StaticRef<SpiHostRegisters>, cpu_clk: u32) -> Self {
+        SpiHost {
+            registers: base,
+            client: OptionalCell::empty(),
+            busy: Cell::new(false),
+            chip_select: Cell::new(0),
+            cpu_clk: Cell::new(cpu_clk),
+            tsclk: Cell::new(0),
+            tx_buf: TakeCell::empty(),
+            rx_buf: TakeCell::empty(),
+            tx_len: Cell::new(0),
+            rx_len: Cell::new(0),
+            tx_offset: Cell::new(0),
+            rx_offset: Cell::new(0),
+        }
+    }
+
+    pub fn handle_interrupt(&self) {
+        let regs = self.registers;
+        let irq = regs.intr_state.extract();
+        self.disable_interrupts();
+
+        if irq.is_set(intr::ERROR) {
+            if self.rx_buf.is_some() {
+                self.rx_buf.take().map(|rx_buf| {
```

Comment:
You don't need to check `is_some()` then use `map()`, just use `map()` directly

Reply:
But you don't even need that, just pass `self.rx_buf.take()` directly to `read_write_done()`